### ISSUE TITLE
Error in reactor loop escaped: closed stream (IOError)

### ISSUE THREAD

**Issue by alaz** [puma/puma#1852]
I am experiencing an issue very similar to #352 under Puma 4.0.1, JRuby 9.2.7 and Rails 5.2.3 (the latest versions).

I have lots of the following traces in the log files:

```
2019-07-18T08:47:16.812765547Z Error in reactor loop escaped: closed stream (IOError)
2019-07-18T08:47:16.812804278Z org/nio4r/Selector.java:139:in `deregister'
2019-07-18T08:47:16.812813845Z /usr/src/app/vendor/bundle/jruby/2.5.0/gems/puma-4.0.1-java/lib/puma/reactor.rb:295:in `clear_monitor'
2019-07-18T08:47:16.812822570Z /usr/src/app/vendor/bundle/jruby/2.5.0/gems/puma-4.0.1-java/lib/puma/reactor.rb:231:in `block in run_internal'
2019-07-18T08:47:16.812831287Z org/jruby/RubyArray.java:1792:in `each'
2019-07-18T08:47:16.812839424Z /usr/src/app/vendor/bundle/jruby/2.5.0/gems/puma-4.0.1-java/lib/puma/reactor.rb:157:in `run_internal'
2019-07-18T08:47:16.812847964Z /usr/src/app/vendor/bundle/jruby/2.5.0/gems/puma-4.0.1-java/lib/puma/reactor.rb:311:in `block in run_in_thread'
```

**Comment by nateberkopec** [comment]
This has been happening in the tests a lot as well.

**Comment by Jesus** [comment]
I believe this is actually a bug in `nio4r`. I've opened an issue in their repo to try and confirm this.

**Comment by Jesus** [comment]
This can be closed, confirmed it was a bug in nio4r and it's already been fixed.

**Comment by nateberkopec** [comment]
So, to clarify for future Googlers: 

If you are seeing, with JRuby,

```
2019-07-18T08:47:16.812765547Z Error in reactor loop escaped: closed stream (IOError)
```

with nio4r 2.4.0 or less, upgrade to nio4r > 2.4.0.

**Comment by shkm** [comment]
> So, to clarify for future Googlers:
> 
> If you are seeing, with JRuby,
> 
> ```
> 2019-07-18T08:47:16.812765547Z Error in reactor loop escaped: closed stream (IOError)
> ```
> 
> with nio4r 2.4.0 or less, upgrade to nio4r > 2.4.0.

Also worth noting that this is the fix with MRI (2.5.5 in this case).

**Comment by nateberkopec** [comment]
This is a super generic error message in Puma that can come from any number of internal causes. The message above was directed at a single, particular cause (JRuby, nio4r < 2.4).

If that doesn't describe your situation, [please submit a new bug report. ](https://github.com/puma/puma/issues/new?assignees=&labels=&template=bug_report.md&title=)
